### ISSUE TITLE
--dry-run doesn't catch malformed schema

### ISSUE THREAD

**Issue by kirs** [kubernetes/kubernetes#42057]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: minikube v0.16.0
- **OS** (e.g. from /etc/os-release): `Darwin Kirs-MacBook-Pro.local 16.1.0 Darwin Kernel Version 16.1.0: Thu Oct 13 21:26:57 PDT 2016; root:xnu-3789.21.3~60/RELEASE_X86_64 x86_64`

**What happened**:

I got a YAML that doesn't follow the kubernetes schema (see the `metadata.labels.name` field):

```
$  cat broken.yml
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: web
  labels:
    name:
      not_a_name:
      - 1
      - 2
    app: basic
spec:
  rules:
  - host: basic.example.com
    http:
      paths:
      - backend:
          serviceName: web
          servicePort: 80
```

Trying to validate the file with dry run:

```
$ kubectl create -f broken.yml --dry-run=true
ingress "web" created (dry run)
```

Looks valid, but then it will fail when trying to apply it:

```
$ kubectl create -f broken.yml
Error from server (BadRequest): error when creating "broken.yml": Ingress in version "v1beta1" cannot be handled as a Ingress: [pos 96]: json: expect char '"' but got char '{'
```

**What you expected to happen**:

It's expected that `--dry-run` would catch the malformed YAML.

**Comment by smarterclayton** [comment]
The error is actually that schema validation is not catching this problem at all, but the server is rejecting the input.  So --dry-run doesn't talk to the server and doesn't report anything.  

SwaggerSchema.ValidateBytes should flag this up front.

**Comment by kirs** [comment]
I'll be happy to send a PR with a fix! Do you mean that we should add a check [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/schema.go#L229) to assert the `labels` type?

**Comment by AdoHe** [comment]
```
SwaggerSchema.ValidateBytes should flag this up front, since map[string]string is explicitly known.
```
@smarterclayton I am afraid such validation is hard to do on client side(at least not elegant), the field type of `metadata.labels` is `object` and now we have no way to validate this.

**Comment by AdoHe** [comment]
and of course if we want to add such validation, some workaround can be achieved.

**Comment by kirs** [comment]
So that you better understand my use case: we have many apps and every single of them has a set of YAMLs with k8s resource definitions. It's a common problem that someone introduces a syntax error (easy to catch with a YAML parser) or a YAML that doesn't comply with k8s schema. We'd like to enforce the projects to have valid k8s resources by validating the configs on CI with `--dry-run`.

Right now there are two issues with `dry-run`:

1) It doesn't detect cases like I described above
2) It requires connection to k8s server and ideally we wouldn't want to give that access to CI jobs

With the validation difficulties mentioned by @AdoHe, my idea is that both issues could be solved by a separate utility that would parse and validate YAMLs against schemas defined in [apimachinery](https://github.com/kubernetes/apimachinery) repo. I'll be happy to work on that.

Do you think that's something that would be valuable for Kubernetes community?

**Comment by smarterclayton** [comment]
The schema for the labels and annotations fields should be more precise - I don't think we need a separate tool, because this is expected to work ootb.

Does open API have a type for map string / string?  We can always use a format.

**Comment by AdoHe** [comment]
```
The schema for the labels and annotations fields should be more precise
```
I agree. will take a look at OpenAPI more specifically.

**Comment by pwittrock** [comment]
@apelisse related to validation

**Comment by apelisse** [comment]
> Does open API have a type for map string / string?

Yes, and that's what it is:

      "labels": {
      "description": "Map of string keys and values that can be used to organize and categorize (scope and select) objects. May match selectors of replication controllers and services. More info: http://kubernetes.io/docs/user-guide/labels",
      "type": "object",
      "additionalProperties": {
       "type": "string"
      }

Literally means that it should enforce `map[string]string`.

I'm working on doing this validation now, and expect this to work OOTB.

**Comment by kirs** [comment]
Looks like bug is fixed (running on kubernetes v1.9.0):

```
$ kubectl apply --dry-run -f broken.yml
error: error validating "broken.yml": error validating data: ValidationError(Ingress.metadata.labels.name): invalid type for io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta.labels: got "map", expected "string"; if you choose to ignore these errors, turn validation off with --validate=false
```
